### PR TITLE
Azure Storage File Service

### PR DESCRIPTION
**Pull request by sima-zhu**

### CODE REVIEW

File: core/azure-core/src/main/java/com/azure/core/http/rest/VoidResponse.java [hunk 1]
```diff
@@ -20,4 +20,8 @@ public VoidResponse(HttpRequest request, int statusCode, HttpHeaders headers) {
         super(request, statusCode, headers, null);
     }
 
+    public VoidResponse(Response response) {
```
Comment (jianghaolu) [inline core/azure-core/src/main/java/com/azure/core/http/rest/VoidResponse.java]: docs

File: storage/client/src/main/java/com/azure/storage/file/models/FileRef.java [hunk 2]
```diff
@@ -4,4 +4,38 @@
...
+        this.name = name;
+        this.isDirectory = isDirectory;
+        this.fileProperty = fileProperty;
+    }
+
+    public String name() {
+        return name;
+    }
+
+    public void name(final String name) {
```
Comment (jianghaolu) [inline storage/client/src/main/java/com/azure/storage/file/models/FileRef.java]: Setters should return the class type for fluent usage.

**Comment by sima-zhu** [comment]
Tests here are only testing the basic functionality. Will continue working on tests.

File: storage/client/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java [hunk 3]
```diff
@@ -3,77 +3,331 @@
...
+         return client.directorys().deleteWithRestResponseAsync(shareName, directoryName, Context.NONE).map(VoidResponse::new)
+            .map(VoidResponse::new);
     }
 
+    /**
+     * Get Properties
+     * @param shareSnapshot
+     * @return
+     */
     public Mono<Response<DirectoryProperties>> getProperties(String shareSnapshot) {
```
Comment (alzimmermsft) [inline storage/client/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java]: Should have a getProperties that doesn't take a snapshot

File: storage/client/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java [hunk 4]
```diff
@@ -3,77 +3,331 @@
...
-    public Flux<HandleItem> getHandles(int maxResult, boolean recursive) {
-        throw new UnsupportedOperationException();
+    /**
+     * Get open handles on a directory or a file
+     * @param maxResult
+     * @param recursive
+     * @return
+     */
+    public Flux<HandleItem> getHandles(Integer maxResult, boolean recursive) {
+        final Integer resultCount = (maxResult == null) ? MAX_RETURN_ITEMS: maxResult;
```
Comment (alzimmermsft) [inline storage/client/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java]: Again should have checks that the service should handle

**Comment by sima-zhu** [comment]
Next step: Update to the right test json file.

File: storage/client/src/main/java/com/azure/storage/blob/AppendBlobClientBuilder.java [hunk 5]
```diff
@@ -3,7 +3,7 @@
 
 package com.azure.storage.blob;
 
-import com.azure.core.configuration.Configuration;
+import com.azure.core.util.configuration.Configuration;
```
Comment (JonathanGiles) [inline storage/client/src/main/java/com/azure/storage/blob/AppendBlobClientBuilder.java]: I'm not sure this PR should be making changes to the storage blob files?

File: storage/client/src/main/java/com/azure/storage/common/credentials/SharedKeyCredential.java [hunk 6]
```diff
@@ -89,17 +89,17 @@ private String buildStringToSign(URL requestURL, String httpMethod, Map<String,
...
-            headers.getOrDefault("If-Modified-Since", ""),
-            headers.getOrDefault("If-Match", ""),
-            headers.getOrDefault("If-None-Match", ""),
-            headers.getOrDefault("If-Unmodified-Since", ""),
-            headers.getOrDefault("Range", ""),
+            getStandardHeaderValue(headers, "If-Modified-Since"),
+            getStandardHeaderValue(headers, "If-Match"),
+            getStandardHeaderValue(headers, "If-None-Match"),
+            getStandardHeaderValue(headers, "If-Unmodified-Since"),
+            getStandardHeaderValue(headers, "Range"),
```
Comment (JonathanGiles) [inline storage/client/src/main/java/com/azure/storage/common/credentials/SharedKeyCredential.java]: Was this change necessary?

File: storage/client/src/main/java/com/azure/storage/common/credentials/package-info.java [hunk 7]
```diff
@@ -2,6 +2,6 @@
 // Licensed under the MIT License.
 
 /**
- * This package contains credentials used by Azure Storage services.
+ * This package contains credential used by Azure Storage services.
```
Comment (JonathanGiles) [inline storage/client/src/main/java/com/azure/storage/common/credentials/package-info.java]: This change should be reverted.

File: storage/client/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java [hunk 8]
```diff
@@ -3,77 +3,527 @@
...
+ * and list or force close handles of the directory or file.
+ *
+ * <p><strong>Instantiating an Asynchronous Directory Client</strong></p>
+ *
+ * <pre>
+ * DirectoryAsyncClient client = DirectoryAsyncClient.builder()
+ *     .connectionString(connectionString)
+ *     .endpoint(endpoint)
+ *     .buildAsync();
+ * </pre>
```
Comment (JonathanGiles) [inline storage/client/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java]: This sample code needs to be extracted out into a sample file.

File: storage/client/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java [hunk 9]
```diff
@@ -3,77 +3,527 @@
...
+import java.util.List;
+import java.util.Map;
 import reactor.core.publisher.Flux;
 import reactor.core.publisher.Mono;
 
-import java.util.Map;
-
+/**
+ * This class provides a client that contains all the operations for interacting with directory in Azure Storage File Service.
+ * Operations allowed by the client are creating, deleting and listing subdirectory and file, retrieving properties, , setting metadata
```
Comment (JonathanGiles) [inline storage/client/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java]:
```suggestion
 * Operations allowed by the client are creating, deleting and listing subdirectory and file, retrieving properties, setting metadata
```

File: storage/client/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java [hunk 10]
```diff
@@ -3,77 +3,527 @@
...
+     */
+    public FileAsyncClient getFileClient(String fileName) {
+        String filePath = directoryName + "/" + fileName;
+        return new FileAsyncClient(azureFileStorageClient, shareName, filePath, null);
     }
 
+    /**
+     * Constructs a DirectoryAsyncClient that interacts with the specified directory.
+     *
+     * <p>If the file doesn't exist in the storage account {@link DirectoryAsyncClient#create()} create} in the client will
```
Comment (JonathanGiles) [inline storage/client/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java]: s/file/directory

File: storage/client/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java [hunk 11]
```diff
@@ -3,77 +3,527 @@
...
+     *
+     * <p>Retrieve directory properties</p>
+     *
+     * <pre>
+     * client.getProperties()
+     *    .subscribe(response -&gt; {
+     *        DirectoryProperties properties = response.value();
+     *        System.out.printf("Directory latest modified date is %s.", properties.lastModified());
+     *    });
+     * </pre>
```
Comment (JonathanGiles) [inline storage/client/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java]: Make into a code snippet

File: storage/client/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java [hunk 12]
```diff
@@ -3,77 +3,527 @@
...
-    public Flux<FileRef> listFilesAndDirectories(String prefix, int maxResults, String shareSnapshot) {
-        throw new UnsupportedOperationException();
+    /**
+     * Lists all shares in the storage account with their prefix or snapshots.
+     *
+     * <p><strong>Code Samples</strong></p>
+     *
+     * <p>List all directories with "subdir" prefix and return 10 results in the account</p>
+     *
+     * @codesnippet com.azure.storage.file.directoryAsyncClient.listFilesAndDirectories
```
Comment (JonathanGiles) [inline storage/client/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java]: Generate the JavaDoc and review that the code snippet is visible in all places. I suspect this won't show up as expected.

File: storage/client/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java [hunk 13]
```diff
@@ -3,77 +3,527 @@
...
+     *              System.out.printf("Close %d handles.", numOfClosedHandles)
+     *     )});
+     * </pre>
+     * @param handleId Specifies the handle ID to be closed. Use an asterisk ('*') as a wildcard string to specify all handles.
+     * @param recursive A boolean value that specifies if the operation should also apply to the files and subdirectories of the directory specified in the URI.
+     * @return The counts of number of handles closed
+     */
     public Flux<Integer> forceCloseHandles(String handleId, boolean recursive) {
-        throw new UnsupportedOperationException();
+        return azureFileStorageClient.directorys().forceCloseHandlesWithRestResponseAsync(shareName, directoryName, handleId, null, null, shareSnapshot, recursive, Context.NONE)
```
Comment (JonathanGiles) [inline storage/client/src/main/java/com/azure/storage/file/DirectoryAsyncClient.java]: directories, not directorys. I see a few API with this typo, please ensure it is all resolved and spelt correctly.

**Comment by sima-zhu** [comment]
Close this one as we moved to another PR
